packer/linux/conf/bin/bk-install-elastic-stack.sh

```diff
@@ -37,16 +37,46 @@ PLUGINS_ENABLED=()
 [[ $DOCKER_LOGIN_PLUGIN_ENABLED == "true" ]] && PLUGINS_ENABLED+=("docker-login")
 
 # cfn-env is sourced by the environment hook in builds
-cat << EOF > /var/lib/buildkite-agent/cfn-env
-export DOCKER_VERSION=$DOCKER_VERSION
-export BUILDKITE_STACK_NAME=$BUILDKITE_STACK_NAME
-export BUILDKITE_STACK_VERSION=$BUILDKITE_STACK_VERSION
-export BUILDKITE_AGENTS_PER_INSTANCE=$BUILDKITE_AGENTS_PER_INSTANCE
-export BUILDKITE_SECRETS_BUCKET=$BUILDKITE_SECRETS_BUCKET
-export AWS_DEFAULT_REGION=$AWS_REGION
-export AWS_REGION=$AWS_REGION
-export PLUGINS_ENABLED="${PLUGINS_ENABLED[*]-}"
-export BUILDKITE_ECR_POLICY=${BUILDKITE_ECR_POLICY:-none}
+
+# We will create it in two steps so that we don't need to go crazy with quoting and escaping. The
+# first sets up a helper function, the second populates the default values for some environment
+# variables.
+
+# Step 1: Helper function.  Note that we clobber the target file and DO NOT apply variable
+# substitution, this is controlled by the double-quoted "EOF".
+cat <<- "EOF" > /var/lib/buildkite-agent/cfn-env
+	# The Buildkite agent sets a number of variables such as AWS_DEFAULT_REGION to fixed values which
+	# are determined at AMI-build-time.  However, sometimes a user might want to override such variables
+	# using an env: block in their pipeline.yml.  This little helper is sets the environment variables
+	# buildkite-agent and plugins expect, except if a user want to override them, for example to do a
+	# deployment to a region other than where the Buildkite agent lives.
+	function set_unless_present() {
+	    local target=$1
+	    local value=$2
+
+	    if [[ -v "${target}" ]]; then
+	        echo "^^^ +++"
+	        echo "⚠️ ${target} already set, NOT overriding! (current value: \"${!target}\")"
```

Comment:
```suggestion
	        echo "⚠️ ${target} modified by user, NOT modifying! (current value: \"${!target}\")"
```

I'm not 100% sure what's clearest here.  Perhaps you folks can shed light on what would be most obvious to the user.

Reply:
I think I like the "set" and "overriding" language here.

I would like to include _where_ it has already been set from. We can’t deduce where it’s from with 100% certainty because by the time the bootstrap and by extension the environment hook runs it could be inherited from the agent’s process environment or set by the Buildkite Step / Job environment as in your case 🤔 Maybe we just need to call out both:

```suggestion
	        echo "⚠️ ${target} already set, NOT overriding! (current value \"${!target}\", set by Buildkite Step env configuration or inherited from the buildkite-agent process environment)"
```

Reply:
Yeah, the uncertainty of "where this value comes from" is why i wanted input 👍 Will update the copy!